src/decompress.c

```diff
@@ -106,22 +239,360 @@ void LoadCompressedSpritePaletteOverrideBuffer(const struct CompressedSpritePale
+{
+    freqs[15] = 0;
+    for (u32 i = 0; i < 3; i++)
+    {
+        for (u32 j = 0; j < 5; j++)
+        {
+            freqs[i*5 + j] = (packedFreqs[i] >> (6*j)) & PACKED_FREQ_MASK;
+        }
+        freqs[15] += (packedFreqs[i] & PARTIAL_FREQ_MASK) >> (30 - 2*i);
+    }
+}
+
+void DecompressDataVram(const u32 *src, void *dest)
+{
+    struct CompressionHeader header;
+    CpuCopy32(src, &header, 8);
+    switch (header.mode)
+    {
+        case MODE_LZ77:
+            LZ77UnCompVram(src, dest);
+            break;
+        default:
+            SmolDecompressData(&header, &src[2], dest);
+    }
+}
+
+void DecompressDataWram(const u32 *src, void *dest)
+{
+    struct CompressionHeader header;
+    CpuCopy32(src, &header, 8);
+    switch (header.mode)
+    {
+        case MODE_LZ77:
+            LZ77UnCompWram(src, dest);
+            break;
+        default:
+            SmolDecompressData(&header, &src[2], dest);
+    }
+}
+
+void BuildDecompressionTable(u32 *packedFreqs, struct DecodeYK *table, u8 *symbolTable)
+{
+    u8 *freqs = Alloc(16);
+    u32 currCol = 0;
+    UnpackFrequencies(packedFreqs, freqs);
+    for (u32 i = 0; i < 16; i++)
+    {
+        if (freqs[i] != 0)
+        {
+            //  Copying 16-bit sized data from EWRAM to array, IMPROVEMENT POSSIBLE
+            memcpy(&table[currCol], &ykTemplate[freqs[i]], freqs[i]*sizeof(struct DecodeYK));
+            //  Setting 8-bit array elements to a single symbol, IMPROVEMENT POSSIBLE
+            for (u32 j = 0; j < freqs[i]; j++)
+                symbolTable[currCol + j] = i;
+            currCol += freqs[i];
+        }
+    }
+    Free(freqs);
+}
+
+void DecodeLOtANS(const u32 *data, u32 *readIndex, u32 *bitIndex, struct DecodeYK *ykTable, u8 *symbolTable, u8 *resultVec, u32 *state, u32 count)
+{
+    u32 currBits = data[*readIndex];
+    for (u32 currSym = 0; currSym < count; currSym++)
+    {
+        u8 symbol = 0;
+        for (u32 currNibble = 0; currNibble < 2; currNibble++)
+        {
+            symbol += symbolTable[*state] << (currNibble*4);
+            u16 currK = ykTable[*state].kVal;
+            u16 nextState = ykTable[*state].yVal << currK;
+            nextState += (currBits >> *bitIndex) & ((1u << currK)-1);
+            if (*bitIndex + currK < 32)
+            {
+                *bitIndex += currK;
+            }
+            else if (*bitIndex + currK == 32)
+            {
+                *readIndex += 1;
+                currBits = data[*readIndex];
+                *bitIndex = 0;
+            }
+            else if ((*bitIndex + currK) > 32)
+            {
+                *readIndex += 1;
+                currBits = data[*readIndex];
+                u32 remainder = *bitIndex + currK - 32;
+                nextState += (data[*readIndex] & ((1u << remainder) - 1)) << (currK - remainder);
+                *bitIndex = remainder;
+            }
+            *state = nextState-64;
+        }
+        resultVec[currSym] = symbol;
+    }
+}
```

Comment:
This is where the tANS magic happens.
And where the biggest question marks are.